### ISSUE TITLE
Object Detection  SUN RGB-D

### ISSUE THREAD

**Issue by beatrriello-hue** [PowerShell/vscode-powershell#5369]
### Prerequisites

- [x] I have written a descriptive issue title.
- [x] I have searched all [_open and closed_ issues](https://github.com/PowerShell/vscode-powershell/issues?q=is%3Aissue) to ensure it has not already been reported.
- [x] I have read the [troubleshooting](https://github.com/PowerShell/vscode-powershell/blob/main/docs/troubleshooting.md) guide.
- [x] I am sure this issue is with the _extension itself_ and does not reproduce in a standalone [PowerShell](https://github.com/PowerShell/PowerShell/issues/new/choose) instance.
- [x] I have verified that I am using the latest version of Visual Studio Code and the PowerShell extension.
- [x] If this is a security issue, I have read the [security issue reporting guidance](https://github.com/PowerShell/vscode-powershell/blob/main/SECURITY.md).

### Summary

This research addresses the challenges of 2D object detec-
tion in cluttered indoor environments by proposing an en-
hanced Single Shot MultiBox Detector (SSD) tailored for
the SUNRGBD dataset. Standard SSD architectures often
struggle with the extreme scale variance and severe oc-
clusions typical of indoor scenes; to mitigate these limita-
tions, we implement a Feature Pyramid Network (FPN) in-
tegration. This architectural refinement utilizes a top-down
pathway with nearest-neighbor interpolation to fuse high-
level semantic information from the extra convolutional lay-
ers with high-resolution features from the VGG16 backbone
(conv4_3 and conv7). To ensure training stability on the het-
erogeneous and often noisy SUNRGBD metadata, a robust
"SUNRobust" data engineering pipeline was developed to
filter degenerate bounding boxes and corrupted JSON en-
tries. Furthermore, we replace the standard Cross-Entropy
loss with a hybrid MultiBox Loss that incorporates Focal
Loss (γ = 2.0, α = 0.5) combined with Hard Negative
Mining at a 3:1 ratio to prioritize difficult spatial samples.
Experimental results demonstrate that the synergy between
hierarchical feature fusion and robust data filtering leads
to a significant improvement in mean Average Precision
(mAP), particularly for small-scale categories like ’lamp’
and ’monitor’, while maintaining real-time inference capa-
bilities on a single GPU

### PowerShell Version

```console
Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      5.1.26100.7462                                                                          
PSEdition                      Desktop                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion                   10.0.26100.7462                                                                         
CLRVersion                     4.0.30319.42000                                                                         
WSManStackVersion              3.0                                                                                     
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1
```

### Visual Studio Code Version

```console
1.108.0
x64
```

### Extension Version

```console
ms-vscode.powershell@2025.4.0
```

### Steps to Reproduce

To ensure the scientific validity of our results, we established a rigorous four-phase pipeline. This process covers everything from raw data acquisition to final model evaluation.

5.1 Phase 1: Data Preparation and Flattening
The original SUN RGB-D dataset is distributed in a complex nested directory structure that is incompatible with standard PyTorch DataLoaders. We developed a custom script to centralize these resources.

Acquisition: Download the official SUNRGBD.zip from the Princeton portal.

Flattening: Run the flatten_sunrgbd.py script. This script traverses the hierarchy and centralizes 10,335 images and their corresponding JSON annotations into a unified structure.

Indexing: The script automatically generates train.txt and test.txt, containing the unique IDs required by the SUNDetection class for deterministic data splitting.

5.2 Phase 2: Environment Configuration
The project is built on Python 3.11 and requires CUDA-enabled hardware for efficient training.

Core Dependencies: Install PyTorch, Torchvision, OpenCV, and NumPy.

Geometry Setup: In data/config.py, verify that the 'sun' configuration dictionary correctly specifies the 8,732 anchor boxes (priors) and the normalization variances (0.1, 0.2).

Weights Initialization: Download the pre-trained VGG16 weights (vgg16_reducedfc.pth) and place them in the weights/ directory to serve as the backbone initialization.

5.3 Phase 3: Robust Training Pipeline
The training is executed via train_clean.py, which integrates our defensive data layer.

Execution: Start the training by running the command: python train_clean.py.

Robustness Mechanism: Our custom SUNRobust class monitors the data stream. If it encounters a malformed JSON or a "degenerate" bounding box (where width or height is zero), it automatically skips the sample and fetches the next valid one. This prevents the training from crashing during the 1,200 iterations.

Optimization: The script utilizes Automatic Mixed Precision (AMP) via torch.amp.autocast. This allows the model to fit a batch size of 8 into memory while maintaining high-speed gradient updates.

Checkpointing: Model weights are saved every 100 iterations as .pth files in the weights/ folder for subsequent evaluation.

5.4 Phase 4: Inference and Evaluation
Once training is complete, the model's performance can be assessed qualitatively and quantitatively.

Inference: To test on new images, run infer.py specifying the model path and the target image. The script applies the FPN fusion, runs the forward pass, and performs Non-Maximum Suppression (NMS) with a threshold of 0.45.

Metric Calculation: Run eval.py to compute the mean Average Precision (mAP) across the 14 indoor categories. The script generates a detections.pkl file containing all prediction scores for rigorous statistical analysis.
